modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java

```diff
@@ -413,17 +301,30 @@ private void onLwmChanged(ChangeLowWatermarkEventParameters parameters) {
         inBusyLock(busyLock, () -> {
             int lwmCatalogVersion = catalogService.activeCatalogVersion(parameters.newLowWatermark().longValue());
 
-            Iterator<ReadOnlyIndexInfo> it = readOnlyIndexes.iterator();
+            readOnlyIndexes.removeIf(readOnlyIndexInfo -> readOnlyIndexInfo.indexRemovalCatalogVersion() <= lwmCatalogVersion);
+        });
+    }
 
-            while (it.hasNext()) {
-                ReadOnlyIndexInfo readOnlyIndexInfo = it.next();
+    private static long activationTs(IndexMeta indexMeta, MetaIndexStatus status) {
+        return indexMeta.statusChange(status).activationTimestamp();
+    }
 
-                if (readOnlyIndexInfo.indexRemovalCatalogVersion() <= lwmCatalogVersion) {
-                    it.remove();
+    private static ReadOnlyIndexInfo toReadOnlyIndexInfo(IndexMeta indexMeta) {
+        assert indexMeta.status() == MetaIndexStatus.READ_ONLY : "indexId=" + indexMeta.indexId() + ", status=" + indexMeta.status();
 
-                    tableVersionByIndexId.remove(readOnlyIndexInfo.indexId());
-                }
-            }
-        });
+        long activationTs;
+
+        if (indexMeta.statusChanges().containsKey(MetaIndexStatus.STOPPING)) {
+            activationTs = activationTs(indexMeta, MetaIndexStatus.STOPPING);
+        } else {
+            activationTs = activationTs(indexMeta, MetaIndexStatus.READ_ONLY);
```

Comment:
Let's add Nullable to indexMeta.statusChange()? javadoc - "Returns the index status change, {@code null} if absent"

Reply:
No need, there is method **org.apache.ignite.internal.table.distributed.index.IndexMeta#statusChangeNullable** for this.

Reply:
Didn't notice the exception - then javadoc for statusChange() is incorrect, about null if absent))